File: programs/pob/src/instructions/close_round.rs
use anchor_lang::prelude::*;

use crate::error::PobError;
use crate::state::{Config, RoundCommitment};

#[derive(Accounts)]
pub struct CloseRound<'info> {
    #[account(mut)]
    pub beneficiary: Signer<'info>,

    /// The config account
    #[account(
        seeds = [Config::SEED_PREFIX.as_ref()],
        bump = config.bump
    )]
    pub config: Account<'info, Config>,

    /// The round commitment account
    #[account(
        mut,
        close = beneficiary,
        seeds = [
            RoundCommitment::SEED_PREFIX.as_ref(),
            &round_commitment.seed,
        ],
        bump = round_commitment.bump
    )]
    pub round_commitment: Box<Account<'info, RoundCommitment>>,
}

pub fn handler(ctx: Context<CloseRound>) -> Result<()> {
    let current_slot = Clock::get()?.slot;
    let round_commitment = &ctx.accounts.round_commitment;
    let config = &ctx.accounts.config;

    // Check if round has ended and grace period has elapsed
    require!(
        current_slot > round_commitment.end_slot,
        PobError::RoundNotActive
    );
    
    require!(
        current_slot > round_commitment.end_slot.saturating_add(config.round_close_grace_slots),
        PobError::RoundGracePeriodNotElapsed
    );

    Ok(())
}

File: cli/commands/pob/close_round.ts
import { connectPob, getFlag, submitTx } from '../../shared/cli-utils'
import { getPobConfigPda, getRoundCommitmentPda } from '../../../sdk/pda/pob'
import { PublicKey } from '@solana/web3.js'

async function main() {
  const { wallet, pobProgram, connection } = await connectPob()

  // Get round from seed or PDA
  const seedHex = getFlag('--seed')
  const roundPdaStr = getFlag('--round')

  let roundPda: PublicKey
  if (roundPdaStr) {
    roundPda = new PublicKey(roundPdaStr)
  } else if (seedHex) {
    const seed = Buffer.from(seedHex, 'hex')
    if (seed.length !== 32) {
      throw new Error('Seed must be 32 bytes (64 hex characters)')
    }
    ;[roundPda] = getRoundCommitmentPda(pobProgram, seed)
  } else {
    throw new Error('Either --seed or --round must be provided')
  }

  const [configPda] = getPobConfigPda(pobProgram)

  console.log('Closing challenge round...')
  console.log({
    roundPda: roundPda.toBase58(),
    beneficiary: wallet.publicKey.toBase58(),
  })

  // Check if grace period has elapsed
  const round = await pobProgram.account.roundCommitment.fetch(roundPda)
  const config = await pobProgram.account.config.fetch(configPda)
  const currentSlot = await connection.getSlot()
  const graceEndSlot =
    round.endSlot.toNumber() + config.roundCloseGraceSlots.toNumber()

  console.log({
    currentSlot,
    endSlot: round.endSlot.toNumber(),
    graceSlots: config.roundCloseGraceSlots.toNumber(),
    graceEndSlot,
  })

  if (currentSlot <= graceEndSlot) {
    console.warn(
      `⚠️  Warning: Grace period not elapsed yet. Need to wait until slot ${graceEndSlot}`,
    )
  }

  const itx = await pobProgram.methods
    .closeRound()
    .accountsPartial({
      beneficiary: wallet.publicKey,
      config: configPda,
      roundCommitment: roundPda,
    })
    .instruction()

  await submitTx(connection, wallet, itx)
  console.log('✅ Round closed successfully')
  console.log('Rent returned to:', wallet.publicKey.toBase58())
}

main().catch(console.error)
